libraries/chain_kv/include/b1/session/cache.hpp

```diff
@@ -0,0 +1,438 @@
+   void             clear();
+
+   template <typename Iterable>
+   const std::pair<std::vector<key_value>, std::unordered_set<shared_bytes>> read(const Iterable& keys) const;
+
+   template <typename Iterable>
+   void write(const Iterable& key_values);
+
+   template <typename Iterable>
+   void erase(const Iterable& keys);
+
+   template <typename Data_store, typename Iterable>
+   void write_to(Data_store& ds, const Iterable& keys) const;
+
+   template <typename Data_store, typename Iterable>
+   void read_from(const Data_store& ds, const Iterable& keys);
+
+   iterator       find(const shared_bytes& key);
+   const_iterator find(const shared_bytes& key) const;
+   iterator       begin();
+   const_iterator begin() const;
+   iterator       end();
+   const_iterator end() const;
+   iterator       lower_bound(const shared_bytes& key);
+   const_iterator lower_bound(const shared_bytes& key) const;
+   iterator       upper_bound(const shared_bytes& key);
+   const_iterator upper_bound(const shared_bytes& key) const;
+
+   const std::shared_ptr<Allocator>&      memory_allocator();
+   const std::shared_ptr<const Allocator> memory_allocator() const;
+
+ private:
+   template <typename On_cache_hit, typename On_cache_miss>
+   const key_value& find_(const shared_bytes& key, const On_cache_hit& cache_hit,
+                          const On_cache_miss& cache_miss) const;
+
+ private:
+   std::shared_ptr<Allocator> m_allocator;
+   cache_type                 m_cache;
+};
+
+template <typename Allocator>
+cache<Allocator> make_cache(std::shared_ptr<Allocator> a) {
+   return { std::move(a) };
+}
+
+template <typename Allocator>
+cache<Allocator>::cache(std::shared_ptr<Allocator> memory_allocator) : m_allocator{ std::move(memory_allocator) } {}
+
+template <typename Allocator>
+const std::shared_ptr<Allocator>& cache<Allocator>::memory_allocator() {
+   return m_allocator;
+}
+
+template <typename Allocator>
+const std::shared_ptr<const Allocator> cache<Allocator>::memory_allocator() const {
+   return m_allocator;
+}
+
+// Searches the cache for the given key.
+//
+// \tparam on_cache_hit A functor to invoke when the key is found.  It has the following signature
+// void(cache_type::iterator& it) \tparam on_cache_miss A functor to invoke when the key is not found.  It has the
+// following singature void(void). \param key The key to search the cache for. \param cache_hit The functor to invoke
+// when the key is found. \param cache_miss The functor to invoke when the key is not found. \return A reference to the
+// key_value instance if found, key_value::invalid otherwise.
+template <typename Allocator>
+template <typename on_cache_hit, typename on_cache_miss>
+const key_value& cache<Allocator>::find_(const shared_bytes& key, const on_cache_hit& cache_hit,
+                                         const on_cache_miss& cache_miss) const {
+   auto it = m_cache.find(key);
+
+   if (it == std::end(m_cache)) {
```

Comment:
Yes.  The argument about it being more characters to type is highly dependent on the naming of the data member, I could have easily called `m_cache` something else that made it shorter to type.  I prefer the usage of `std::begin`, `std::end` because it is generic and  I could change m_cache to be a type that doesn't have that interface on it and not be bogged down in changing the `.end()` calls to `std::end`.